src/foam/nanos/auth/ServiceProvider.js

```diff
@@ -34,5 +49,102 @@ foam.CLASS({
           getId().substring(0, 1).toUpperCase() + getId().substring(1) + " Service Provider Capability";
       `
     }
+  ],
+
+  methods: [
+    {
+      name: 'setupSpid', 
+      args: [
+        { name: 'x', javaType: 'foam.core.X' },
+        { name: 'user', javaType: 'foam.nanos.auth.User' }
+      ],
+      documentation: `
+        Method responsible for setting up a user's ServiceProvider capability by finding
+        the prerequisites of the ServiceProvider and granting those along with it
+      `,
+      javaCode: `           
+        Logger logger = (Logger) x.get("logger");
+        DAO userCapabilityJunctionDAO = (DAO) x.get("bareUserCapabilityJunctionDAO");
+        CrunchService crunchService = (CrunchService) x.get("crunchService");
+
+        // get grantPath of the service provider capability
+        List<Capability> grantPath = (List<Capability>) crunchService.getCapabilityPath(x, getId(), false);
+
+        try {
+          // for each capability in the grantPath of the spid capability,
+          // find the ucj and update its status to granted, or create a ucj none found
+          
+          UserCapabilityJunction ucj;
+          for ( Capability capability : grantPath ) {
+            ucj = (UserCapabilityJunction) userCapabilityJunctionDAO.find(AND(
+              EQ(UserCapabilityJunction.SOURCE_ID, user.getId()),
+              EQ(UserCapabilityJunction.TARGET_ID, capability.getId())
+            ));
+            if ( ucj != null ) 
+              ucj = (UserCapabilityJunction) ucj.fclone();
+            else 
+              ucj = new UserCapabilityJunction.Builder(x).setSourceId(user.getId()).setTargetId(capability.getId()).build();
+            ucj.setStatus(CapabilityJunctionStatus.GRANTED);
+            ucj = (UserCapabilityJunction) userCapabilityJunctionDAO.put_(x, ucj);
+            if ( ucj == null || ucj.getStatus() !=foam.nanos.crunch.CapabilityJunctionStatus.GRANTED )
+              throw new RuntimeException("Error setting up UserCapabilityJunction for user: " + user.getId() + " and spid: " + getId());
+          }
+        } catch (Exception e) {
+          logger.warning(e);
+        }
+      `
+    },
+    {
+      name: 'removeSpid',
+      args: [
+        { name: 'x', javaType: 'foam.core.X' },
+        { name: 'user', javaType: 'foam.nanos.auth.User' }
+      ],
+      documentation: `
+        Method used for removing a user's old ServiceProvider capability and any capability that is a part
+        of that capability through prerequisite junctions.
+        Called before a user is assigned a new ServiceProvider capability
+      `,
+      javaCode: `
+        CrunchService crunchService = (CrunchService) x.get("crunchService");
+        DAO userCapabilityJunctionDAO = (DAO) x.get("bareUserCapabilityJunctionDAO");
+
+        // find list of old spids to remove from user
+
+        List<UserCapabilityJunction> spidsToRemove = new ArrayList<UserCapabilityJunction>();    
```

Comment:
can be removed, as it will get re-assigned on line 123.

Reply:
thanks, done